# Ball Falling due To Gravity

Time step for all simulations is dt=0.01 because any slower or faster would cause delays in the simulation, or lag that affected the visuals

In [1]:
from vpython import *

<IPython.core.display.Javascript object>

In [2]:
def ballFallingDueToGravity(height):
    scene = canvas()
    ball = sphere(pos=vec(0,height,0), radius=0.3, color=color.yellow, make_trail = True)
    velocity = 0
    velocityList = [0]
    secondsList = [0]
    g = 9.8
    dt = 0.01
    seconds = 0
    maxv = 0
    
    while ball.pos.y > 0:
        rate(100)
        seconds = seconds + 1
        velocity = velocity - g * dt
        
        if velocity < maxv:
            maxv = velocity
            
        # decrease position
        ball.pos.y = ball.pos.y + velocity * dt
        secondsList.append(seconds/100)
        velocityList.append(velocity)

    print(height,"m Drop Time: ",seconds/100," seconds")
    return secondsList, velocityList, maxv

In [3]:
x = ballFallingDueToGravity(100)

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

100 m Drop Time:  4.52  seconds


The simulation uses a rate of 100 which means the simulation is refreshed 100 times per second. Therefore the final seconds count needs to be divided by 100 to get the final time.

### Calculations

v = integral(a dt) = a * t
x = integral(v dt) = integral(a*t dt) = 0.5*a*t^2
t = (2x/a)^1/2

a = 9.8
x = 100
t = (200/(9.8))^1/2
t = (20.41)^1/2
t = 4.518s

### Results

The time to drop the ball 100m from our simulation was 4.52 seconds. The calculations we made to verify our results gave a value of 4.518s. Therefore, we can conclude that the simulation accurately represents a ball dropping in free fall.  

# Ball Bouncing

In [4]:
from vpython import *

In [5]:
scene = canvas()
ball = sphere(pos=vec(0,10,0), radius=0.3, color=color.red, make_trail = True)
ball.velocity = 0
ground = box(pos=vector(0,0,0), size=vector(15,0.2,15), color=color.green)

<IPython.core.display.Javascript object>

In [6]:
dt = 0.01
g = 9.8 #m/s^2
x = 0

# manually end if simulation is running longer than needed
while x<700:
    rate(100)
    x = x+1
    ball.pos.y = ball.pos.y + ball.velocity*dt
    if ball.pos.y < ball.radius:
        # send ball in opposite direction (no energy lost assuming)
        ball.velocity = abs(ball.velocity)
    else:
        ball.velocity = ball.velocity - g*dt

### Assumptions

I assumed that the simulation would be using elastic collision meaning no energy is lost upon collision. Therefore, the ball bounces at the same height infinitely. Mass of the ball is negligible in this circumstance.

# Ball Launched from Cannon

In [7]:
from vpython import *

In [8]:
def launchFromCannon(angle):
    scene = canvas()
    ball = sphere(pos=vector(0,0,0), radius=6, color=color.red, make_trail = True)
    ground = box(pos=vector(0,0,0), size=vector(1000,1,30), color=color.green)

    seconds = 0
    v0 = 40  # asssuming cannon launches at 40 m/s?
    g = 9.8  #m/s^2
    dt = 0.01 
    theta = angle*(3.1415926/180) #angle from degrees to radians
    velocity = vector(v0*cos(theta),v0*sin(theta),0)

    # same as before essentially
    while ball.pos.y >= 0:  
        rate(100) 
        seconds = seconds + 1
        #decrease position
        ball.pos = ball.pos + velocity * dt 
        velocity.y = velocity.y - g * dt  
    return seconds/100, ball.pos.x

In [9]:
time,meters = launchFromCannon(60) #angle in degrees
print("Launch Time in Seconds: ",time)
print("Meters Travelled Y Axis: ",meters)

<IPython.core.display.Javascript object>

Launch Time in Seconds:  7.08
Meters Travelled Y Axis:  141.6000043811165


In [ ]:
time,meters = launchFromCannon(20) #angle in degrees
print("Launch Time in Seconds: ",time)
print("Meters Travelled Y Axis: ",meters)

<IPython.core.display.Javascript object>

In [ ]:
time,meters = launchFromCannon(45) #angle in degrees
print("Launch Time in Seconds: ",time)
print("Meters Travelled Y Axis: ",meters)

### Calculations

The equation for launching a ball from a cannon is (Vertical/Y Component):
y(t) = v0*sin(theta)*t-(0.5*g*t^2) (theta=launch angle)

The equation for launching a ball from a cannon is (Horizontal/X component):
x(t) = v0*cos(θ)t

The Y component is used for the Y component of the velocity vector and the X component is used for the X component of the velocity vector.

We can pick a number to assume the cannon launches the ball at and I picked 40 m/s which is our v0. Then we pick an angle to shoot the ball from.

Calculating Time:
At t=0, y=0
0 = v0*sin(theta)*t-(0.5*g*t^2)
0 = t(v0*sin(theta)-(0.5*g*t^2)
t=0

0 = v0*sin(theta)-(0.5*g*t^2)
0.5*g*t^2 = v0*sin(theta)
t = sqrt(v0*sin(theta) / (0.5*g))
t = sqrt(40*sin(0.785398) / 4.9
t = 5.7722993114 seconds
##### Verified by the simulation above, some rounding errors in hand calculation

Calculating Distance Traveled X:
x(t) = v0*cos(theta)t
x(t) = 40*cos(0.785398)*5.7722993114
x(t) = 163.265297315 m
##### Verified by the simulation above, some rounding errors in hand calculation

### Results

The smaller the angle, the smaller the horizontal distance traveled and the shorter the launch time.

Comparing the calculations made by hand and the simulation made, we can infer that the simulation is an accurate representation of a ball being launched from a cannon. The mass of the ball is negligible in this situation.

# Graphs

This section includes the graphs for the ball in freefall and the ball bouncing

## Ball Dropping Due To Gravity

### Velocity Over Time Without Air Resistance

In [ ]:
from vpython import *
import matplotlib.pyplot as plt
velocityList = []
secondsList = []

In [ ]:
def graphHeightDrops(secondsList,velocityList,height):
    plt.plot(secondsList,velocityList)
    plt.plot(secondsList, velocityList, label=f"{height} Drop")
    plt.xlabel('Seconds')
    plt.ylabel('Velocity')
    plt.title('Velocity Over Time Without Air Resistance')
    plt.legend()

In [ ]:
secondsList, velocityList, maxv = ballFallingDueToGravity(100)
graphHeightDrops(secondsList,velocityList,"100")
print("100m Max Velocity: ",maxv)

secondsList, velocityList, maxv = ballFallingDueToGravity(50)
graphHeightDrops(secondsList,velocityList,"50")
print("50m Max Velocity: ",maxv)

secondsList, velocityList, maxv = ballFallingDueToGravity(25)
graphHeightDrops(secondsList,velocityList,"25")
print("25m Max Velocity: ",maxv)

plt.show()

### Velocity Over Time With Air Resistance

In [ ]:
from vpython import *

In [ ]:
def ballFallingDueToGravityWithAirResistance(height):
    scene = canvas()
    ball = sphere(pos=vec(0,height,0), radius=0.3, color=color.yellow, make_trail = True)
    velocity = 0
    velocityList = [0]
    secondsList = [0]
    g = 9.8
    dt = 0.01
    seconds = 0
    t = 0

    # air resistance
    rho = 1.225  # kg/m^3
    C = 0.47  # drag coefficient
    r = 0.3  # radius
    A = 3.1415926*r*r  # pir^2
    m = 0.2  # assuming mass of ball is 0.2 kg?
    k = (1/2)*C*rho*A  # air resistance constant
    maxv = 0
    
    while ball.pos.y > 0:
        rate(100)

        #air pull ball down
        drag = -k*velocity*abs(velocity)

        if velocity > maxv:
            maxv = velocity
        
        a = g + drag
        velocity = velocity + a*dt
        ball.pos.y -= velocity * dt
        
        seconds = seconds + 1
        secondsList.append(seconds/100)
        velocityList.append(velocity)

    print(height,"m Drop Time: ",seconds/100," seconds")
    return secondsList, velocityList, maxv

In [ ]:
import matplotlib.pyplot as plt
def graphHeightDropsAirResist(secondsList,velocityList,height):
    plt.plot(secondsList,velocityList)
    plt.plot(secondsList, velocityList, label=f"{height} Drop")
    plt.xlabel('Seconds')
    plt.ylabel('Velocity')
    plt.title('Velocity Over Time With Air Resistance')
    plt.legend()

In [ ]:
secondsList, velocityList, maxv = ballFallingDueToGravityWithAirResistance(100)
graphHeightDropsAirResist(secondsList,velocityList,"100")
print("100m Max Velocity: ",maxv)

secondsList, velocityList, maxv = ballFallingDueToGravityWithAirResistance(50)
graphHeightDropsAirResist(secondsList,velocityList,"50")
print("50m Max Velocity: ",maxv)

secondsList, velocityList, maxv = ballFallingDueToGravityWithAirResistance(25)
graphHeightDropsAirResist(secondsList,velocityList,"25")
print("25m Max Velocity: ",maxv)

plt.show()

### Results

Regardless of drop the height the ball is dropped from, it appears that the ball will have similar velocities when in free fall. The freefall velocity without air resistance is a linear line which makes sense considering there are no forces opposing the gravity pulling the ball down. The max velocity for each simulation decreases with a decrease in height the ball is dropped from. This makes sense as the ball has less time to accelerate before it reaches the ground.

The velocity of the freefall with air resistance causes a parabola in the velocity over time graph. This makes sense as the force pushing against gravity on the ball would result in a peak velocity of some value. In this case it appears to be about 10 m/s for all three simulations.

In this simulation I assumed the ball would have a mass of 0.2 kg and a radius of 0.3 m.

# Ball Bouncing

## Position Over Time With Elastic Collision

In [ ]:
from vpython import *
import matplotlib.pyplot as plt

In [ ]:
def ElasticCollision(height):
    scene = canvas()
    ball = sphere(pos=vec(0,height,0), radius=0.3, color=color.red, make_trail = True)
    ball.velocity = 0
    ground = box(pos=vector(0,0,0), size=vector(15,0.2,15), color=color.green)
    dt = 0.01
    g = 9.8 #m/s^2
    x = 0
    heightlist = [height]
    secondslist = [0]
    seconds = 0
    
    # manually end if simulation is running longer than needed
    while x<1000:
        rate(100)
        x = x+1
        seconds = seconds + 1
        ball.pos.y = ball.pos.y + ball.velocity*dt
        if ball.pos.y < ball.radius:
            ball.velocity = abs(ball.velocity)
        else:
            ball.velocity = ball.velocity - g*dt
    
        heightlist.append(ball.pos.y)
        secondslist.append(seconds)
    
    return secondslist, heightlist

In [ ]:
def graphElasticCollision(secondsList,heightlist,height):
    plt.plot(secondsList,heightlist)
    plt.plot(secondsList, heightlist, label=f"{height} Drop")
    plt.xlabel('Seconds')
    plt.ylabel('Height')
    plt.title('Position Over Time: Elastic Collision')
    plt.legend()

In [ ]:
secondsList, heightlist = ElasticCollision(10)
graphElasticCollision(secondsList,heightlist,"10")

secondsList, heightlist = ElasticCollision(5)
graphElasticCollision(secondsList,heightlist,"5")

secondsList, heightlist = ElasticCollision(2)
graphElasticCollision(secondsList,heightlist,"2")

plt.show()

## Velocity Over Time With Inelastic Collision

In [ ]:
from vpython import *
import matplotlib.pyplot as plt

In [ ]:
def InelasticCollision(height):
    scene = canvas()
    ball = sphere(pos=vec(0,height,0), radius=0.3, color=color.red, make_trail = True)
    ball.velocity = 0
    ground = box(pos=vector(0,-0.1,0), size=vector(15,0.2,15), color=color.green)
    dt = 0.01
    g = -9.8 #m/s^2
    x = 0
    heightlist = [height]
    secondslist = [0]
    seconds = 0
    coe = 0.99
    if height >=5:
        coe = 0.99
    else:
        coe = 0.85
    
    # manually end if simulation is running longer than needed
    while x<2000:
        rate(100)
        seconds = seconds + 1
        x = x+1
        ball.velocity = ball.velocity + g*dt
        ball.pos.y = ball.pos.y + ball.velocity*dt

       # print(ball.velocity)
        
        if ball.pos.y <= ball.radius:
            ball.velocity = -ball.velocity * coe
            if ball.pos.y < 0.01 and abs(ball.velocity) < 0.001:
                break  

        if ball.pos.y <= 0:
            break
    
        heightlist.append(ball.pos.y)
        secondslist.append(seconds)
    
    return secondslist, heightlist

In [ ]:
def graphInelasticCollision(secondsList,heightlist,height):
    plt.plot(secondsList,heightlist)
    plt.plot(secondsList, heightlist, label=f"{height} Drop")
    plt.xlabel('Seconds')
    plt.ylabel('Height')
    plt.title('Position Over Time: Inelastic Collision')
    plt.legend()

In [ ]:
secondsList, heightlist = InelasticCollision(10)
graphInelasticCollision(secondsList,heightlist,"10")

secondsList, heightlist = InelasticCollision(5)
graphInelasticCollision(secondsList,heightlist,"5")

secondsList, heightlist = InelasticCollision(2)
graphInelasticCollision(secondsList,heightlist,"2")

plt.show()

### Results

The graphs for both the inelastic and elastic collisions make sense. In the elastic collision, no energy is lost upon collision. The ball would essentially bounce forever, and thus the position over time remains steady parabolas. The smaller the height the ball is dropped from, the higher the frequency of the graph.

The inelastic collision graph shows that as the frequency of the graph increases, the position reaches a steady state where the ball can be shown as not moving. 